staging/src/kubesphere.io/utils/helm/executor.go

```diff
@@ -36,41 +56,496 @@ type Executor interface {
+cat > ./.local-helm-output.yaml
+kustomize build
+`
+)
+
+var (
+	errorTimedOutToWaitResource = errors.New("timed out waiting for resources to be ready")
+)
+
 type executor struct {
+	// target cluster client
+	client.Client
+	kubeConfig string
+	namespace  string
+	// helm release name
+	releaseName string
+	// helm action Config
+	helmConf  *action.Configuration
+	helmImage string
+	// add labels to helm chart
+	labels map[string]string
+	// add annotations to helm chart
+	annotations map[string]string
+	dryRun      bool
 }
 
 type Option func(*executor)
 
+// SetDryRun sets the dryRun option.
+func SetDryRun(dryRun bool) Option {
+	return func(e *executor) {
+		e.dryRun = dryRun
+	}
+}
+
+// SetAnnotations sets extra annotations added to all resources in chart.
+func SetAnnotations(annotations map[string]string) Option {
+	return func(e *executor) {
+		e.annotations = annotations
+	}
+}
+
+// SetLabels sets extra labels added to all resources in chart.
+func SetLabels(labels map[string]string) Option {
+	return func(e *executor) {
+		e.labels = labels
+	}
+}
+
+// SetHelmImage sets the helmImage option.
+func SetHelmImage(helmImage string) Option {
+	return func(e *executor) {
+		e.helmImage = helmImage
+	}
+}
+
+// SetKubeConfig sets the kube config data of the target cluster.
+func SetKubeConfig(kubeConfig string) Option {
+	return func(e *executor) {
+		e.kubeConfig = kubeConfig
+	}
+}
+
 // NewExecutor generates a new Executor instance with the following parameters:
-//   - kubeConfig: kube config data of the target cluster
 //   - namespace: the namespace of the helm release
 //   - releaseName: the helm release name
 //   - options: functions to set optional parameters
-func NewExecutor(kubeConfig, namespace, releaseName string, options ...Option) (Executor, error) {
-	return &executor{}, nil
+func NewExecutor(namespace, releaseName string, options ...Option) (Executor, error) {
+	e := &executor{
+		namespace:   namespace,
+		releaseName: releaseName,
+		helmImage:   "kubesphere/helm:latest",
+	}
+	for _, option := range options {
+		option(e)
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig([]byte(e.kubeConfig))
+	if err != nil {
+		return nil, err
+	}
+	clusterClient, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+	e.Client = clusterClient
+
+	klog.V(8).Infof("namespace: %s, release name: %s, kube config:%s", e.namespace, e.releaseName, e.kubeConfig)
+
+	getter := NewClusterRESTClientGetter(e.kubeConfig, e.namespace)
+	e.helmConf = new(action.Configuration)
+	if err = e.helmConf.Init(getter, e.namespace, "", klog.Infof); err != nil {
+		return nil, err
+	}
+
+	if err = e.ensureNamespaceAndServiceAccount(context.Background(), e.namespace); err != nil {
+		return nil, fmt.Errorf("ensureNamespaceAndServiceAccount failed: %v", err)
+	}
+	return e, nil
+}
+
+func (e *executor) ensureNamespaceAndServiceAccount(ctx context.Context, chartNamespace string) error {
```

Comment:
This behavior seems too obscure, cluster role and cluster role binding must create by the end user. You can add the `--create-namespace` option for the executor.

Reply:
I don't need to create a namespace, but mainly I need a `ServiceAccount` for the `Job` that can create `Pod` permissions, and if I ask the user to provide it beforehand, the library is too difficult to use, rancher did the same thing: https://github.com/k3s-io/helm-controller/blob/8ea81e4fb70861878ab3a912d8be0a7d21337d49/pkg/controllers/chart/chart.go#L408-L444